tools/ir-generator/methods.cpp

```diff
@@ -56,6 +56,45 @@ const ordered_map<cstring, IrMethod::info_t> IrMethod::Generate = {
         buf << ";" << std::endl;
         buf << cl->indent << "}";
         return buf.str(); } } },
+{ "equiv", { &NamedType::Bool,
+             { new IrField(new ReferenceType(new NamedType(IrClass::nodeClass), true), "a_") },
+             CONST + IN_IMPL + CONCRETE_ONLY + OVERRIDE,
+    [](IrClass *cl, Util::SourceInfo, cstring) -> cstring {
+        std::stringstream buf;
+        buf << "{" << std::endl;
+        buf << cl->indent << cl->indent
+            << "if (static_cast<const Node *>(this) == &a_) return true;\n";
+        buf << cl->indent << cl->indent << "if (typeid(*this) != typeid(a_)) return false;\n";
+        if (auto parent = cl->getParent()) {
+            while (parent && parent->kind == NodeKind::Abstract)
+                parent = parent->getParent();
+            if (parent && parent->name != "Node")
+                buf << cl->indent << cl->indent << "if (!" << parent->name
+                    << "::equiv(a_)) return false;\n"; }
+        bool first = true;
+        for (auto f : *cl->getFields()) {
+            if (*f->type == NamedType::SourceInfo) continue;  // FIXME -- deal with SourcInfo
```

Comment:
I don't know if there is anything to fix here; I suspect that you really want the equivalence to ignore the position.